procrastinate/exceptions.py

```diff
@@ -45,8 +45,11 @@ class JobRetry(ProcrastinateException):
     Job should be retried.
     """
 
-    def __init__(self, scheduled_at: datetime.datetime):
+    def __init__(
```

Comment:
I think it might make more sense to attach the retry decision to the exception, even if the should_retry will always be True.

Reply:
Wouldn't that mean we could break something? Isn't it public API (it has some docstring but is, on the other hand, not in the API reference)? We also would have to calculate `schedule_at` somewhere else (`RetryDecision` uses `schedule_in`).

Reply:
I don't think it's exposed anywhere. Do you have an idea how a user might manipulate a `JobRetry`? I could be missing something (honnestly, I really could) but I don't _think_ it's exposed at the moment.

> We also would have to calculate schedule_at somewhere else (RetryDecision uses schedule_in).

Hm, I haven't dived into the question enough to see exactly what you have in mind, but... `schedule_in` and `schedule_at` are often slightly annoying in the API, but it's meant to be as helpful as possible for the user. `schedule_at` is the source of truth, `schedule_in` is the syntactic sugar. I guess this points towards the fact `RetryDecision` should probably have a `schedule_at` attribute, and the constructor should accept both. Probably worth factoring [this code](https://github.com/procrastinate-org/procrastinate/blob/f349c5be9044efb7276a6fa375f2e05ad2868150/procrastinate/tasks.py#L46-L55) out to a function and use it in both places, WDYT?

Reply:
> I don't think it's exposed anywhere. Do you have an idea how a user might manipulate a `JobRetry`? I could be missing something (honnestly, I really could) but I don't _think_ it's exposed at the moment.

What does finally end up in the Reference API? I thought everything with a docstring, but that doesn't seem to be the case.

> 
> > We also would have to calculate schedule_at somewhere else (RetryDecision uses schedule_in).
> 
> Hm, I haven't dived into the question enough to see exactly what you have in mind, but... `schedule_in` and `schedule_at` are often slightly annoying in the API, but it's meant to be as helpful as possible for the user. `schedule_at` is the source of truth, `schedule_in` is the syntactic sugar. I guess this points towards the fact `RetryDecision` should probably have a `schedule_at` attribute, and the constructor should accept both. 

From a user perspective, I find `schedule_in` more convenient because most of the time, I just want to do something like "just retry in 10 minutes", regardless of the current date or time (and let the library figure it out). In the rare cases where I would like to specify a specific date/time, I can still do that with a bit of calculation.
I am not a big fan of having mutually exclusive options like `schedule_in` and `schedule_at`, especially as one can be quite easily calculated from the other. If it's ok for you I would stay with `schedule_in`, but would rename it to `retry_in` in the `RetryDecision`. I then would move the logic to conversion into `retry_at` from [`retry.py`](https://github.com/procrastinate-org/procrastinate/blob/234f5f0ef3571cd9ff9053ca174c98e18d0f6631/procrastinate/retry.py#L29) to somewhere else (as we only store the `RetryDecision` in the `JobRetry`).
Is this ok for you, or do you really want to have both `retry_in` and `retry_at` in `RetryDescision`. I can live with that, too 😂.

> Probably worth factoring [this code](https://github.com/procrastinate-org/procrastinate/blob/f349c5be9044efb7276a6fa375f2e05ad2868150/procrastinate/tasks.py#L46-L55) out to a function and use it in both places, WDYT?

I see task scheduling and retries as two separate features even though they both ultimately use the same underlying logic (the `schedule_at` column in Postgres). At least I wouldn't refactor that in this issue. I also think that the API there is okay and not worth breaking.

Reply:
Hmm, I just found out that `schedule_in` of the task is more advanced than I thought (especially regarding `TimeDeltaParams`). Now, I am a bit insecure about what the best decision would be if we really want to make both APIs similar. Then maybe we should indeed allow `retry_in` and `retry_at` in the `RetryDecision` (even when both are mutually exclusive, which I kind of dislike a bit). You decide 😉.

Reply:
> What does finally end up in the Reference API? I thought everything with a docstring, but that doesn't seem to be the case.

I try to put a docstring in anything public, but the opposite isn't true. Anything people should need to import (including in typing) should be `procrastinate.__all__` (in `__init__.py`) and should be documented in `reference.rst` (thus the docstrings). This is all manual, I try to maintain a mental model of what's officially exposed or not but I don't think such a thing appears explicitly in the code apart from what I mentioned. It could be interesting to make hard rules.

> [...retry_in, retry_at]

Of course for the naming, go ahead. Regarding the mutually exclusive: do you know that thanks to `typing.overload`, you can make it explicit that the 2 params can't be used together (for the type checker)? (also... Maybe this ship has sailed, but I should probably have done a single param that receives either a `datetime` or a `timedelta`).

> If it's ok for you I would stay with schedule_in, but would rename it to retry_in in the RetryDecision. I then would move the logic to conversion into retry_at from [retry.py](https://github.com/procrastinate-org/procrastinate/blob/234f5f0ef3571cd9ff9053ca174c98e18d0f6631/procrastinate/retry.py#L29) to somewhere else (as we only store the RetryDecision in the JobRetry).

Could also be in the `RetryDecision` constructor? But elsewhere works too (in that case, `RetryDecision` will keep 2 attributes). I think it's not very important, but I'd love to only have this logic coded once and reused. Datetimes are hard enough, if we can avoid consistency issues :D

> You decide 😉.

You drive a hard bargain. Ok, if I'm in charge, let's say we keep both, for the sake of API consistency throughout Procrastinate. If we can do it when we configure a task, we can do it at retry time too. Might be overkill but at least no one will wonder why it's available on one side and not on the other. But let me be clear: it's a subjective decision. I don't foresee this decision making us sad in the future (compared to a _bunch_ of past decisons making me feel sad in the present :D )

Reply:
> > What does finally end up in the Reference API? I thought everything with a docstring, but that doesn't seem to be the case.
> 
> I try to put a docstring in anything public, but the opposite isn't true. Anything people should need to import (including in typing) should be `procrastinate.__all__` (in `__init__.py`) and should be documented in `reference.rst` (thus the docstrings). This is all manual, I try to maintain a mental model of what's officially exposed or not but I don't think such a thing appears explicitly in the code apart from what I mentioned. It could be interesting to make hard rules.

Thanks, good to know.

> 
> > [...retry_in, retry_at]
> 
> Of course for the naming, go ahead. Regarding the mutually exclusive: do you know that thanks to `typing.overload`, you can make it explicit that the 2 params can't be used together (for the type checker)? (also... Maybe this ship has sailed, but I should probably have done a single param that receives either a `datetime` or a `timedelta`).
> 
> > If it's ok for you I would stay with schedule_in, but would rename it to retry_in in the RetryDecision. I then would move the logic to conversion into retry_at from [retry.py](https://github.com/procrastinate-org/procrastinate/blob/234f5f0ef3571cd9ff9053ca174c98e18d0f6631/procrastinate/retry.py#L29) to somewhere else (as we only store the RetryDecision in the JobRetry).
> 
> Could also be in the `RetryDecision` constructor? But elsewhere works too (in that case, `RetryDecision` will keep 2 attributes). I think it's not very important, but I'd love to only have this logic coded once and reused. Datetimes are hard enough, if we can avoid consistency issues :D

Yes, doing this in the constructor and keeping only `retry_at` sounds nice. Thereby, I could kill two birds with one stone by also using `typing.overload` on the constructor.

> > You decide 😉.
> 
> You drive a hard bargain. Ok, if I'm in charge, let's say we keep both, for the sake of API consistency throughout Procrastinate. If we can do it when we configure a task, we can do it at retry time too. Might be overkill but at least no one will wonder why it's available on one side and not on the other. But let me be clear: it's a subjective decision. I don't foresee this decision making us sad in the future (compared to a _bunch_ of past decisons making me feel sad in the present :D )

I already edited my post, and it's nice to see that we came to the same conclusion 😉

Reply:
> I already edited my post, and it's nice to see that we came to the same conclusion 😉

Very nice indeed. Have I mentioned I'm thrilled to have you on board ? Because I am 😀